### ISSUE TITLE
[libbeat] pipeline.Close() throwing panic

### ISSUE THREAD

**Issue by fearful-symmetry** [elastic/beats#13892]
This came about as a result of my work on a docker plugin.

When a container stops and the libbeat pipeline assocated with it closes, libbeat throws a kernel panic. (from the docker logs, cleaned up best I can)

```
panic: send on closed channel
goroutine 89 [running]:
github.com/elastic/beats/libbeat/publisher/pipeline.(*retryer).loop(0xc000291080)
    /go/src/github.com/elastic/beats/libbeat/publisher/pipeline/retry.go:131 
created by github.com/elastic/beats/libbeat/publisher/pipeline.newRetryer
    /go/src/github.com/elastic/beats/libbeat/publisher/pipeline/retry.go:86 +0x11e
```

It's some kind of race, since it doesn't happen 100% of the time. I haven't been able to hack around it. Here's what's happening from the beat side in the close method:

```go
	// deincrement the ref count
	hash := cl.pipelineHash
	pm.pipelines[hash].refCount--

	pm.Logger.Infof("Closing Client first from pipelineManager")
        //This closes the client associated with the pipeline
	err := cl.Close()
	if err != nil {
		return errors.Wrap(err, "error closing client")
	}

	if pm.pipelines[hash].refCount < 1 {
		pm.Logger.Infof("Pipeline closing")
                // now close the pipeline
		pm.pipelines[hash].pipeline.Close()

		delete(pm.pipelines, hash)
	}
```

It appears this is happening when a pipeline closes with an event still in the queue. It seems like this should be handled behavior?

**Comment by fearful-symmetry** [comment]
Alright, can confirm this only happens when the output is blocked.

**Comment by fearful-symmetry** [comment]
Going off an earlier convo with @urso , it looks like the issue is here:

https://github.com/elastic/beats/blob/master/libbeat/publisher/pipeline/controller.go#L85

```go 
func (c *outputController) Close() error {
	c.consumer.sigPause()

	if c.out != nil {
		for _, out := range c.out.outputs {
			out.Close()
		}
		close(c.out.workQueue)
	}

	c.consumer.close()
	c.retryer.close()

	return nil
}
```

That `c.out.workQueue` is probably the one that's causing the panic when it's used here: https://github.com/elastic/beats/blob/master/libbeat/publisher/pipeline/retry.go#L173

**Comment by urso** [comment]
From code location it looks like it is the retryer trying to push some failed batch back onto the shared work queue.

Calling close on the workqueue last should reduce the chance of a panic: https://github.com/elastic/beats/blob/master/libbeat/publisher/pipeline/controller.go#L92

Still, retryer.close only signals shutdown. It does not block until it is shutdown. Same for the consumer (although the consumer should be put on hold).

Turning the consumer+retryer close into a blocking call (add a sync.WaitGroup) + defering workqueue close should resolve the issue.

**Comment by fearful-symmetry** [comment]
Fixed in #13922
